src/repr/src/stats2.rs

```diff
@@ -0,0 +1,618 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Persist Stats for non-primitive types.
+//!
+//! For primitive types please see [`mz_persist_types::stats2`].
+
+use std::collections::BTreeMap;
+
+use arrow::array::{BinaryArray, FixedSizeBinaryArray, StringArray};
+use chrono::{DateTime, NaiveTime};
+use dec::OrderedDecimal;
+use mz_persist_types::columnar::FixedSizeCodec;
+use mz_persist_types::stats::bytes::{BytesStats, FixedSizeBytesStats, FixedSizeBytesStatsKind};
+use mz_persist_types::stats::json::{JsonMapElementStats, JsonStats};
+use mz_persist_types::stats::primitive::PrimitiveStats;
+use mz_persist_types::stats::{ColumnStatKinds, ColumnStats, PrimitiveStatsVariants};
+use mz_persist_types::stats2::ColumnarStatsBuilder;
+use ordered_float::OrderedFloat;
+use prost::Message;
+use uuid::Uuid;
+
+use crate::adt::date::Date;
+use crate::adt::datetime::PackedNaiveTime;
+use crate::adt::interval::{Interval, PackedInterval};
+use crate::adt::jsonb::{JsonbPacker, JsonbRef};
+use crate::adt::numeric::{Numeric, PackedNumeric};
+use crate::adt::timestamp::CheckedTimestamp;
+use crate::row::ProtoDatum;
+use crate::{Datum, Row, ScalarType};
+
+/// Returns a `(lower, upper)` bound from the provided [`ColumnStatKinds`], if applicable.
+pub fn col_values<'a>(
+    typ: &ScalarType,
+    stats: &'a ColumnStatKinds,
+) -> (Option<Datum<'a>>, Option<Datum<'a>>) {
+    use PrimitiveStatsVariants::*;
+
+    match (typ, stats) {
+        (ScalarType::Bool, ColumnStatKinds::Primitive(Bool(stats))) => {
+            let map_datum = |val| if val { Datum::True } else { Datum::False };
+            (stats.lower().map(map_datum), stats.upper().map(map_datum))
+        }
+        (ScalarType::PgLegacyChar, ColumnStatKinds::Primitive(U8(stats))) => (
+            stats.lower().map(Datum::UInt8),
+            stats.upper().map(Datum::UInt8),
```

Comment:
Lots of repetition here... I'd consider adding some kind of map fn that applies the same logic to both bounds.